### DISCUSSION TITLE
Docker swarm or kubernetes for multitenant SAAS invoicing app

### DISCUSSION THREAD

**Discussion by weaseldotro** [BretFisher/ama#171] [Q&A]
I have a small invoice app with 3 components: an API backend (single executable written in golang), a frontend (static JS app built with webpack) and postgresql database.

I want my app to use containers and be cloud agnostic, so no vendor lock-in for database, storage, etc. Also, I would prefer to keep everything as simple and automated as possible.

I want to deploy this app as a software as a service (SAAS). A customer would register on my website and I will provision them with an instance of my app. So for each customer I must deploy a backend, a frontend and a database. I don't need hard multi-tenancy, because the customers will only have access to my app, and not the underlying container. I do need to set resource quotas (cpu, memory, maybe disk space) for each customer.

I have been looking into kubernetes, it is obviously very complex and I've had a hard time configuring storage for postgresql because I want my app to be cloud agnostic, so no azure disk, gce disk, etc., so just use the disk space provided by the host.

Then I looked at docker swarm, it looks way more simple and better for what I need, however I haven't found a decent way to separate the instances of the app (except for maybe adding a prefix to the containers' names).

Do you recommend docker swarm, kubernetes or some other solution?

**Comment by BretFisher** [comment]
Great topic. There's a lot going on here, so I'll try to focus on the key issues. 

Kubernetes and Swarm can both do the basics of what you will need, but I don't think that's a top-3 question you should be asking. The real problem is your time. I would focus on reducing the overhead of systems management.

1. The requirement of "cloud agnostic" will put a large burden on your project and require you to be the full-time admin of everything.  Just because you want to be multi-cloud capable, doesn't mean you can't take advantage of a cloud's features. I think a much better approach would be to create an automated solution using Terraform for AWS first. Then as you *need* to use Azure, on-prem,etc. you would create a slightly different Terraform configuration for that scenario.

2. I almost never recommend running your own DB's in the cloud. How will you back them up, provide failover, restore them, monitor them, etc. Cloud solutions are also much more highly tuned then out-of-the-box containers, so that saves you the time of benchmarking and tuning on your own. Something like AWS RDS (or a different clouds equivalent) automates all that, and can be provisioned with Terraform.  

3. If you're just one person, using a cloud's Kubernetes offering is the way to go. You don't want to be managing the control-plane of clusters yourself. Swarm does have reduced burden of management, but no one provides it as a service, so now you're back to doing monthly server patching, updating Docker, doing Swarm backups, etc.  AWS Fargate for EKS ( Kubernetes )  is the best for automating all of that away.

So if this were my project, I'd likely start with AWS, then create two Terraform configs. One that will provision my networking, EKS + Fargate, load balancers, etc. Then the 2nd would be a single customer's provisioning, creating a RDS database, load balancer rules, and Kubernetes Deployments in their own namespace (ideally no persistent storage requirements for pods, which aren't supported in Fargate). Then I would apply that 2nd Terraform for each client into my single solution that was provided by the first Terraform.

After that worked well, if I wanted Azure/Google/DigitalOcean support, I'd copy the Terraforms and  change them up to work with the database and Kubernetes solution that's provided by that specific cloud.

Good luck!

**Comment by weaseldotro** [comment]
Thanks for your answer.

I do not have any experience with AWS or Azure or Google Cloud, and their prices are much higher than my current provider (Hetzner). Those are the two reasons why I want to use a cloud agnostic deployment.

Also, for my use case the whole point of using Docker Swarm / Kubernetes instead of a DIY hacky solution is "configure it once and then it just works, with minimal or no maintenance". So I would prefer to setup something myself suited for my needs instead of paying a fee for all those features that I will never use included in a provider's offering.
